python/tchannel/io.py

```diff
@@ -19,8 +19,169 @@
+        the stream data once in single direction.
+
+        NOTE: The data which has been read will be discarded in the stream.
+
+        :return:chunk of bytes read from stream
+        """
+        raise NotImplementedError()
+
+    def write(self, chunk):
+        """Async write to internal stream buffer
+
+        :return if stream has been closed, it will raise StreamingException
+        """
+        raise NotImplementedError()
+
+    def close(self):
+        """Close the stream
+
+        It updates the _state to completed.
+
+        :return:
+        """
+        raise NotImplementedError()
+
+
+class InMemArgStream(BaseArgStream):
+
+    def __init__(self, buf=None, auto=True):
+        """In-Memory based arg stream
+
+        :param buf:
+        :return:
+        """
+        super(InMemArgStream, self).__init__()
+        self._stream = buf or ""
+        self._condition = Condition()
+        self.auto = auto
+
+    @tornado.gen.coroutine
+    def read(self):
+        if self._state != StreamState.completed and self._stream == "":
+            yield self._condition.wait()
+
+        chunk = self._stream
+        self._stream = ""
+        raise tornado.gen.Return(chunk)
+
+    @tornado.gen.coroutine
+    def write(self, chunk):
+        if self._state == StreamState.completed:
+            raise StreamingException("Stream has been closed.")
+
+        self._stream += chunk
+        self._condition.notify()
+
+    @tornado.gen.coroutine
+    def close(self):
+        self._state = StreamState.completed
+        self._condition.notify()
+
+
+class PipeArgStream(BaseArgStream):
+
+    def __init__(self, rpipe, wpipe=None, auto=False):
+        """Pipe-based arg stream
+
+        NOTE: if you want to pass a file descriptor of a file, you have to
+        use os.open to get the file descriptor instead of python's open.
+
+        :param rpipe: an integer file descriptor which supports read ops
+        :param wpipe: an integer file descriptor which supports write ops
+        :param auto: flag to indicate to close the stream automatically or not
+        """
+        super(PipeArgStream, self).__init__()
+        self._rpipe = rpipe
+        self._wpipe = wpipe
+
+        self._rs = (PipeIOStream(self._rpipe) if
+                    self._rpipe is not None else None)
+        self._ws = (PipeIOStream(self._wpipe) if
+                    self._wpipe is not None else None)
+        self.auto = auto
+
+    @tornado.gen.coroutine
+    def read(self):
+        if self._state == StreamState.completed or self._rpipe is None:
+            raise tornado.gen.Return("")
+
+        chunk = ""
+        try:
+            chunk = yield self._rs.read_bytes(
+                common.MAX_PAYLOAD_SIZE, partial=True)
+
+        except StreamClosedError:
+            # reach the end of the pipe stream
+            self._state = StreamState.completed
+        finally:
+            raise tornado.gen.Return(chunk)
```

Comment:
Return the value from inside the try block. If you want an empty string returned for StreamClosedError, have a separate `raise Return` inside the `except`. This will return an empty string in case of other exceptions and the system will end up ignoring the actual exceptions.